numba/cpython/unsafe/tuple.py

```diff
@@ -2,11 +2,88 @@
 This file provides internal compiler utilities that support certain special
 operations with tuple and workarounds for limitations enforced in userland.
 """
+import inspect
 
-from numba.core.cgutils import alloca_once
+from numba.core import types, typing
+from numba.core.errors import RequireLiteralValue
 from numba.core.extending import intrinsic
 
 
+# QUESTION: Is there a good place for this function somewhere else?
+#           I keep writing boiler-plate type checks and having a function that
+#           does this for most applications would get rid of much of the
+#           boiler-plate.
+def validate_arg_type(arg, arg_name, valid_types, none_allowed=False,
+                      msg_prefix=None):
+    """Checks that a parameter of an overload has the correct type and prints
+    appropriate error message if not.
+
+    Usage: validate_arg_type(tup, 'tup', types.BaseTuple,
+                             msg_prefix='tuple_setitem: ')
+    """
+    if not isinstance(valid_types, tuple):
+        valid_types = (valid_types, )
+
+    if msg_prefix is None:
+        msg_prefix = ''
+
+    # determine the type name to show to the user
+    name_mapping = {
+        types.Boolean: 'boolean',
+        types.BaseTuple: 'tuple',
+        types.IntegerLiteral: 'constant (literal) integer'
+    }
+
+    valid_names = [None] * len(valid_types)
+
+    # any defined name mapping takes precedence
+    valid_names = [name_mapping.get(type_, name) if name is None
+                   else name
+                   for type_, name in zip(valid_types, valid_names)]
+    # if a type instance is defined, it's name takes precendence
+    valid_names = [getattr(type_, 'name', None) if name is None
+                   else name
+                   for type_, name in zip(valid_types, valid_names)]
+    # fallback to type name
+    valid_names = [getattr(type_, '__name__', None) if name is None
+                   else name
+                   for type_, name in zip(valid_types, valid_names)]
+    for x in valid_names:
+        assert x is not None, f"Cannot find name for type {x}"
```

Comment:
Possibly clearer as:
```suggestion
    def name_for_type(type_):
        # any defined name mapping takes precedence
        name = name_mapping.get(type_)
        if name is not None:
            return name
        # if a type instance is defined, it's name takes precendence
        name = getattr(type_, 'name', None)
        if name is not None:
            return name
        # fallback to type name
        name = getattr(type_, '__name__', None)
        if name is not None:
            return name
        raise ValueError(f"Cannot find name for type {x}")

    valid_names = [name_for_type(type_)  for type_ in valid_types]
```

Reply:
@eric-wieser Thanks for the suggestion.